application/controllers/OptinController.php

```diff
@@ -27,53 +27,183 @@ class OptinController extends LSYii_Controller
+        if (!isset($languageCode) || $languageCode == "" || !$languageCode) {
+            $baseLanguage = $survey->language;
+        } else {
+            $baseLanguage = sanitize_languagecode($languageCode);
+        }
+
+        Yii::app()->setLanguage($baseLanguage);
+
+        if (empty($survey) || !$survey->hasTokensTable) {
+            throw new CHttpException(404, "This survey does not seem to exist. It may have been deleted or the link you were given is outdated or incorrect.");
+        } else {
+            LimeExpressionManager::singleton()->loadTokenInformation($surveyId, $accessToken, false);
+            $token = Token::model($surveyId)->findByAttributes(['token' => $accessToken]);
+
+            if (!isset($token)) {
+                $message = gT('You are not a participant of this survey.');
+            } else {
+                if ($token->emailstatus != 'OptOut') {
+                    $message = gT('You are already a participant of this survey.');
+                } else {
+                    $message = "<p>" . gT('Please confirm that you want to be added back to this survey by clicking the button below.') . '<br>' . gT("After confirmation you may start receiving invitations and reminders for this survey.") . "</p>";
+                    $message .= '<p><a href="' . Yii::app()->createUrl('optin/addtokens', ['surveyid' => $surveyId, 'langcode' => $baseLanguage, 'token' => $accessToken]) . '" class="btn btn-default btn-lg">' . gT("I confirm") . '</a><p>';
+                }
+            }
+        }
+
+        $this->renderHtml($message, $survey);
+    }
+
+    /**
+     * Display the confirmation for global opt in
+     */
+    public function actionparticipants()
+    {
+        Yii::app()->loadHelper('database');
+        Yii::app()->loadHelper('sanitize');
+
+        $surveyId = Yii::app()->request->getQuery('surveyid');
+        $languageCode = Yii::app()->request->getQuery('langcode');
+        $accessToken = Token::sanitizeToken(Yii::app()->request->getQuery('token'));
+
+        //IF there is no survey id, redirect back to the default public page
+        if (!$surveyId) {
+            $this->redirect(['/']);
+        }
+
+        $surveyId = (int) $surveyId; //Make sure it's an integer (protect from SQL injects)
+        $survey = Survey::model()->findByPk($surveyId);
+
+        //Check that there is a SID
+        // Get passed language from form, so that we dont lose this!
+        if (!isset($languageCode) || $languageCode == "" || !$languageCode) {
+            $baseLanguage = $survey->language;
+        } else {
+            $baseLanguage = sanitize_languagecode($languageCode);
+        }
+
+        Yii::app()->setLanguage($baseLanguage);
+
+        if (empty($survey) || !$survey->hasTokensTable) {
+            throw new CHttpException(404, "This survey does not seem to exist. It may have been deleted or the link you were given is outdated or incorrect.");
```

Comment:
Can we register to an notb activated survey ? Or not valid by date survey ?

My opinion : must check active + open 

https://github.com/LimeSurvey/LimeSurvey/blob/14990d854b7ba2298a04965b74264c0fad6377d7/application/models/Survey.php#L457-L458

Reply:
But if it's a global optin : need a surveyid ? Even without …

Reply:
@gabrieljenik : today we can optin/optout on a inactive or date out survey (closed by date) ? I can check but if you already done.

Reply:
> today we can optin/optout on a inactive or date out survey (closed by date)

Yes, we have looked through the code and think so.